doc/03_reference/coverage_plan.rst

```diff
@@ -0,0 +1,310 @@
+
+* **Instr** - A stall caused by a multi-cycle instruction.
+  This can be seen by instructions from categories:
+
+  * **MUL**
+  * **DIV**
+  * **Branch**
+  * **Jump**
+
+* **LdHz** - A load hazard, the instruction in ID/EX depends upon the result of a load that is awaiting a response in writeback.
+  This can be seen by instructions from categories:
+
+  * **ALU**
+  * **Mul**
+  * **Div**
+  * **Branch**
+  * **Jump**
+  * **Load**
+  * **Store**
+  * **CSRAccess**
+
+* **Mem** - Memory stall, the instruction in ID/EX is awaiting a prior memory request to complete before it can begin (to allow precise interrupts on a memory error response). This can be seen for all instruction categories
+
+Privilege Level
+"""""""""""""""
+Ibex can operate at either the M (machine) or U (user) privilege levels.
+Different aspects of the Ibex microarchitecture can be using different privilege levels at once.
+
+* Privilege level of IF stage instruction.
+* Privilege level of ID/EX stage instruction.
+* Privilege level of LSU operation (ID/EX privilege level modified by ``mstatus.mprv`` and ``mstatus.mpp`` settings).
+
+Note that the privilege level of the instruction in WB isn't retained by the microarchitecture and is not relevant to coverage.
+Any instruction that reaches WB can be considered bound to retire and any relevant checks and functionality altered by the privilege mode is dealt with at an earlier stage.
+
+Hazards
+"""""""
+Ibex hazards all occur in the interaction between the ID and EX stage.
+
+* RAW Reg - Read after write hazard, instruction in ID/EX reads a register that writeback is writing.
+  Handled with data forwarding and no stall.
+* RAW Load - Read after write hazard when reading from load destination, a special case of RAW Reg above.
+  Produces a stall (Category LdHz) and shouldn't forward data.
+* RAW Load/Store bytes - Load with bytes overlapping a store immediately before it.
+
+State Specific Behaviour
+""""""""""""""""""""""""
+Some instructions will behave differently depending upon the state of the processor (e.g. the privilege level the instruction executes at, CSR settings or whether the processor is in debug mode).
+
+* Instruction illegal in U Mode (cover execution in U and M mode).
+
+  * ``MRET``
+  * ``WFI``
+  * Access to M-mode CSR.
+
+* Debug mode instructions (cover execution in and out of debug mode).
+
+  * ``DRET``
+  * Access to debug CSRs.
+
+    * ``dcsr``
+    * ``dpc``
+    * ``dscratch0``
+    * ``dscratch1``
+
+  * Access to trigger CSRs (also possible in M mode: cover execution in M mode, debug mode and U mode).
+
+    * ``tselect``
+    * ``tdata1``
+    * ``tdata2``
+    * ``tdata3``
+    * ``mcontext``
+    * ``scontext``
+
+* Loads/stores with ``mstatus.mprv`` set and unset.
+* EBreak behaviour in U/M mode with different ``dcsr.ebreakm`` / ``dcsr.ebreaku`` settings.
+
+Pipeline State
+^^^^^^^^^^^^^^
+Each pipeline stage has some associated state.
+
+* IF stage awaiting fetch, full, or idle.
```

Comment:
I'm going to change this to four categories:
* awaiting fetch
* full and fetching
* full and idle
* empty